counsel.el

```diff
@@ -3254,6 +3254,35 @@ Note: Duplicate elements of `kill-ring' are always deleted."
  '(("d" counsel-yank-pop-action-remove "delete")
    ("r" counsel-yank-pop-action-rotate "rotate")))
 
+;;** `counsel-evil-registers'
+
+(defcustom counsel-evil-registers-height 5
+  "The `ivy-height' of `counsel-evil-registers'."
+  :group 'ivy
+  :type 'integer)
+
+(defun counsel-evil-registers ()
+  "Ivy replacement for `evil-show-registers'."
+  (interactive)
+  (let ((ivy-format-function #'counsel--yank-pop-format-function)
+        (ivy-height counsel-evil-registers-height))
+    (ivy-read "evil-registers: "
+              (cl-loop for (key . val) in (evil-register-list)
+                       collect
+                       (format "[%s]: %s" (char-to-string key)
+                               (or (and val (stringp val) val) "")))
+              :require-match t
+              :action #'counsel-evil-registers-action
+              :caller 'counsel-evil-registers)))
+
+(defun counsel-evil-registers-action (s)
+  "Paste contents of S, trimming the register part.
+
+S will be of the form \"[register]: content\"."
+  (with-ivy-window
+    (insert
+     (string-trim-left s "^\\[.*\\]: "))))
```

Comment:
`string-trim-left` already adds a beginning-of-string (rather than beginning-of-line) anchor to its `REGEXP` argument so you don't have to.

But in general I think it's best to prefer built-in functions over inline ones defined in `subr-x`:

```el
(replace-regexp-in-string "\\`\\[.*\\]: " "" s)
```

Reply:
Hmnn, what is the ` for (not *extremely* familiar with emacs regex.

I do think this particular replacement breaks compared to the first though.

For example, the second (replace-regexp) breaks (ironically) for this.

```
(defun counsel-evil-registers-action (s)
  "Paste contents of S, trimming the register part.

S will be of the form \"[register]: content\"."
  (with-ivy-window
    (insert
     (replace-regexp-in-string "\\`\\[.*\\]: " "" s)
     ;; (string-trim-left s "^\\[.*\\]: ")
     )))
```

As in copying that text and trying to paste it from evil-registers will past an incomplete version.

Reply:
`^` matches the beginning of a string/buffer, but also any character following a newline character within said string/buffer. The backtick only matches the beginning of a string/buffer. The backtick is always preferred unless you're explicitly matching newlines. A rare example where `^` is wrong is when matching filenames, as they can conceivably include actual newline characters as part of their name. See [`(elisp) Regexp Spceial`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Regexp-Special.html) for an explanation of `^` and [`(elisp) Regexp Backslash`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Regexp-Backslash.html) for a description of the backtick.

I think the error might be arising from the use of a greedy wildcard operator. Try this instead:

```el
(replace-regexp-in-string "\\`\\[.*?\\]: " "" s)
```

Reply:
Thanks, that works for me.